File: nexal/channels/channel.py
from __future__ import annotations

import contextlib
from abc import ABC, abstractmethod
from collections.abc import AsyncGenerator
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Callable, Coroutine


OnMessage = Callable[["IncomingMessage"], Coroutine[Any, Any, None]]


@dataclass
class ImageAttachment:
    """An image downloaded from a channel message."""
    data: bytes
    mime_type: str
    filename: str = ""


@dataclass
class IncomingMessage:
    channel: str
    chat_id: str
    sender: str
    text: str
    timestamp: datetime
    is_mentioned: bool = True
    metadata: dict[str, Any] = field(default_factory=dict)
    images: list[ImageAttachment] = field(default_factory=list)
    typing_fn: Callable[[], contextlib.AbstractAsyncContextManager[None]] | None = field(
        default=None, repr=False, compare=False,
    )


class Channel(ABC):
    """Abstract communication channel that can listen for and send messages."""

    @property
    @abstractmethod
    def name(self) -> str: ...

    @property
    def direct_response(self) -> bool:
        """If True, the agent's final text output is sent via ``send()`` automatically.

        Channels that use external skill scripts (Telegram, Discord) return False.
        Channels where the agent's text IS the reply (CLI) return True.
        """
        return False

    @abstractmethod
    async def start(self, on_message: OnMessage) -> None:
        """Start listening for messages. Calls on_message for each incoming message.

        This method should run indefinitely (e.g. polling loop).
        """

    @abstractmethod
    async def send(self, chat_id: str, text: str) -> None:
        """Send a message to the given chat."""

    async def stop(self) -> None:
        """Stop the channel gracefully."""

File: nexal-rs/channel-core/src/channel.rs
//! The [`Channel`] trait — abstract interface for message sources.

use crate::IncomingMessage;

/// A message source (Telegram, Discord, CLI, etc.).
///
/// Channels are long-lived: [`start`](Channel::start) runs until the channel
/// shuts down, calling `on_message` for each incoming message.
#[async_trait::async_trait]
pub trait Channel: Send + Sync + 'static {
    /// Unique identifier for this channel, e.g. `"telegram"`.
    fn name(&self) -> &str;

    /// If `true`, the agent's final text response is automatically sent
    /// back via [`send`](Channel::send).  CLI uses this; Telegram/Discord
    /// instead send messages through skill scripts inside the sandbox.
    fn direct_response(&self) -> bool {
        false
    }

    /// Start listening for messages indefinitely.
    ///
    /// The implementation should call `on_message` for each received message.
    /// This method should block (async) until the channel shuts down.
    async fn start(
        &self,
        on_message: Box<dyn Fn(IncomingMessage) -> futures_like::BoxSendFut + Send + Sync>,
    ) -> anyhow::Result<()>;

    /// Send a text message to a specific chat.
    async fn send(&self, chat_id: &str, text: &str) -> anyhow::Result<()>;

    /// Graceful shutdown.
    async fn stop(&self) -> anyhow::Result<()> {
        Ok(())
    }
}

/// Helper module for the callback type used in [`Channel::start`].
///
/// We avoid pulling in the full `futures` crate by defining a minimal
/// boxed-future type alias.
mod futures_like {
    use std::future::Future;
    use std::pin::Pin;

    /// A boxed, Send future returning `()`.
    pub type BoxSendFut = Pin<Box<dyn Future<Output = ()> + Send>>;
}

// Re-export for use by channel implementations.
#[allow(unused_imports)]
pub use futures_like::BoxSendFut;
